src/ucp/stream/stream_recv.c

```diff
@@ -30,110 +62,302 @@ UCS_PROFILE_FUNC(ucs_status_ptr_t, ucp_stream_recv_data_nb,
-        ucs_list_add_tail(&worker->stream_eps, &ep->ext.stream->list);
-    }
+static UCS_F_ALWAYS_INLINE ucs_status_t
+ucp_stream_am_handler(void *am_arg, void *am_data, size_t am_length,
+                      unsigned am_flags)
+{
+    ucp_worker_h        worker      = am_arg;
+    ucp_stream_am_hdr_t *hdr        = am_data;
+    ucp_ep_ext_stream_t *ep_stream  = NULL;
+    ucp_ep_h            ep;
+    ucp_recv_desc_t     *rdesc;
+    ucp_request_t       *req;
+    ssize_t             unpacked;
 
-    ucs_queue_push(&ep->ext.stream->data, &rdesc->stream_queue);
+    ucs_assert(am_length >= sizeof(ucp_stream_am_hdr_t));
 
-out:
-    UCP_THREAD_CS_EXIT_CONDITIONAL(&worker->mt_lock);
-    return status;
+    rdesc = ucp_stream_am_rdesc_get(worker, am_data, am_length, am_flags);
+    ucs_assert(rdesc != NULL);
 
-err:
-    if (rdesc->flags & UCP_RECV_DESC_FLAG_UCT_DESC) {
-        ucs_mpool_put(rdesc);
+    ep = ucp_stream_am_get_ep(worker, hdr->sender_uuid);
+    if (ucs_unlikely(ep == NULL)) {
+        goto out;
     }
 
-    status = UCS_OK;
-    goto out;
-}
+    ep_stream = ep->ext.stream;
 
-ucs_status_ptr_t ucp_stream_recv_nb(ucp_ep_h ep, void *buffer, size_t *count,
-                                    ucp_datatype_t datatype,
-                                    ucp_stream_recv_callback_t cb,
-                                    unsigned flags)
-{
-    return UCS_STATUS_PTR(UCS_ERR_NOT_IMPLEMENTED);
+    if (ep->flags & UCP_EP_FLAG_STREAM_IS_QUEUED) {
+        ucs_queue_push(&ep_stream->data, &rdesc->stream_queue);
+        goto out;
+    } else if (ucs_queue_is_empty(&ep_stream->reqs)) {
+        ucs_list_add_tail(&worker->stream_eps, &ep_stream->list);
+        ep->flags |= UCP_EP_FLAG_STREAM_IS_QUEUED;
+        ucs_queue_push(&ep_stream->data, &rdesc->stream_queue);
+        goto out;
+    }
+
+    do {
+        req = ucs_queue_head_elem_non_empty(&ep_stream->reqs, ucp_request_t,
+                                            recv.queue);
+        while ((ep_stream->rdesc = ucp_stream_recv_data_nb_internal(ep_stream))) {
+
+            unpacked = ucp_stream_rdata_unpack(ucp_stream_rdesc_data(ep_stream->rdesc),
+                                               ep_stream->rdesc->length, req);
+            if (ucs_unlikely(unpacked < 0)) {
+                goto out;
+            } else if (unpacked < ep_stream->rdesc->length) {
+                ep_stream->rdesc->length -= unpacked;
+                /* This request is full, try next one */
+                ucp_request_complete_stream_recv(req, ep_stream, UCS_OK);
+                break;
+            } else {
+                /* The descriptor is completely processed, go to next */
+                ucs_assert(ep_stream->rdesc->length == unpacked);
+                ucp_stream_rdesc_release(ep_stream->rdesc);
+                ep_stream->rdesc = NULL;
+            }
+        }
+    } while ((ep_stream->rdesc != NULL ) &&
+             !ucs_queue_is_empty(&ep_stream->reqs));
+
```

Comment:
Don't you need to put ep on the worker steam_ep list if ep_stream->rdesc != NULL && ucs_queue_is_empty(&ep_stream->reqs) ? Basically recheck condition at line 320 ?